processor/eqreceipt/app.go

```diff
@@ -0,0 +1,124 @@
+package eqreceipt
+
+import (
+	"context"
+	"encoding/json"
+	"github.com/ONSdigital/census-rm-pubsub-adapter/models/incoming-pubsub"
+	"github.com/ONSdigital/census-rm-pubsub-adapter/models/rabbit"
+	"log"
+
+	"cloud.google.com/go/pubsub"
+	"github.com/pkg/errors"
+	"github.com/streadway/amqp"
+)
+
+type App struct {
+	RabbitConn   *amqp.Connection
+	RabbitChan   *amqp.Channel
+	PubSubClient *pubsub.Client
+	EqRecieptSub *pubsub.Subscription
+	ReceiptChan  chan incoming_pubsub.EqReceipt
+}
+
+func (a *App) Setup(ctx context.Context, rabbitConnectionString, projectID string) {
+
+	//set up rabbit connection
+	var err error
+	a.RabbitConn, err = amqp.Dial(rabbitConnectionString)
+	failOnError(err, "Failed to connect to RabbitMQ")
+
+	a.RabbitChan, err = a.RabbitConn.Channel()
+	failOnError(err, "Failed to open a channel")
+
+	//setup pubsub connection
+	a.PubSubClient, err = pubsub.NewClient(ctx, projectID)
+	failOnError(err, "Pubsub client creation failed")
+	//setup subcriptions
+	a.EqRecieptSub = a.PubSubClient.Subscription("rm-receipt-subscription")
+	a.ReceiptChan = make(chan incoming_pubsub.EqReceipt)
+
+}
+
+func (a *App) Produce(ctx context.Context) {
+	for {
+		select {
+		case eqReceiptReceived := <-a.ReceiptChan:
+			rmMessageToSend, err := convertEqReceiptToRmMessage(&eqReceiptReceived)
+			if err != nil {
+				log.Println(errors.Wrap(err, "failed to convert receipt to message"))
+			}
+			sendRabbitMessage(rmMessageToSend, a.RabbitChan)
+		}
+	}
+
+}
+
+func (a *App) Consume(ctx context.Context) {
+	log.Println("Launched PubSub message listener")
+
+	// Not using the cancel function here (which makes WithCancel a bit redundant!)
+	// Ideally the cancel would be deferred/delegated to a signal watcher to
+	// enable graceful shutdown that can kill the active go routines.
+	cctx, _ := context.WithCancel(ctx)
+	err := a.EqRecieptSub.Receive(cctx, func(ctx context.Context, msg *pubsub.Message) {
+		log.Printf("Got message: %q\n", string(msg.Data))
+
+		eqReceiptReceived := incoming_pubsub.EqReceipt{}
+		json.Unmarshal(msg.Data, &eqReceiptReceived)
+
+		log.Printf("Got QID: %q\n", eqReceiptReceived.Metadata.QuestionnaireID)
+
+		a.ReceiptChan <- eqReceiptReceived
+		msg.Ack()
```

Comment:
I'm afraid you inherited this problem from my first fumbling idiotic foray into the Gopher world. It's really dangerous for us to ack the pubsub message just because we put the EQ receipt onto a channel... we have no idea if the channel consumer managed to successfully send the rabbit message or not.

Fundamentally, if the receiving of PubSub messages is faster than the sending of Rabbit messages, then the channel will start to build up a large amount of 'todo' work, which is only kept in memory. If K8S decided to kill the pubsub pod, then we would lose all those in-flight messages on the channel, because they'd have already been ack'ed, but not yet sent to Rabbit.

We probably need to have a total re-think about this part of the design. Perhaps we could use some kind of bi-directional channel (does such a thing exist @eldeal ?) so that the ack'ing of the PubSub messages isn't done until the Rabbit message has been successfully sent.

Reply:
We've been discussing this, our current nicest idea is to pass the pubsub message into the producer as well so the producer can ack when it is successfully commited to a queue.

Reply:
that sounds like a good idea

Reply:
Have a look at Google's suggested way of doing concurrent PubSub message processing... https://cloud.google.com/pubsub/docs/pull#concurrency_control

Reply:
all in good time, this is a known issue at the mo, i just want to get a basic structure in place for the app and we'll deal with the complexities at a later time, @AdamHawtin and have a few ideas to address this, in the meantime we are listing these issues [here](https://trello.com/c/D7KPvh0q/850-pubsub-adapter)

Reply:
ah hadn't seen that code example, looks quite nice will defo have a look at that cheers @NickGrantONS